lib/routes/builds.js

```diff
@@ -314,28 +319,105 @@ module.exports = function (app) {
    *             type: object
    *     responses:
    *       $ref: '#/responses/Standard'
+   *
+   *   put:
+   *     summary: Puts a fully built npm publish payload structure
+   *     security:
+   *       - basicAuth: []
+   *     produces:
+   *       - "application/json"
+   *     parameters:
+   *       - $ref: '#/parameters/Pkg'
+   *     requestBody:
+   *       description: An 'npm publish'-like JSON data
+   *       content:
+   *         application/json:
+   *           schema:
+   *             type: object
+   *     responses:
+   *       204:
+   *         description: Success
    */
-  app.routes.route('/builds/:pkg')
-    .get(auth, function (req, res, next) {
-      // TODO: fetch the latest build from artifactory or CDN
-      // Investigate how this interacts with the CDN.
-      /* eslint no-unused-vars: 0*/
-      next();
-    })
+  // app.routes.route('/builds/:pkg')
+  //   .get(auth, function (req, res, next) {
+  //     // TODO: fetch the latest build from artifactory or CDN
+  //     // Investigate how this interacts with the CDN.
+  //     /* eslint no-unused-vars: 0*/
+  //     next();
+  //   })
 
-    //
-    // Trigger a build in carpenter through POST request, pipe the response. This
-    // route assumes valid `npm publish` like JSON as data.
-    //
-    .post(auth, function build(req, res, next) {
-      app.carpenter.build({
-        data: req.body || {}
-      }, function client(error, response) {
-        if (error) return next(error);
+  //   //
+  //   // Trigger a build in carpenter through POST request, pipe the response. This
+  //   // route assumes valid `npm publish` like JSON as data.
+  //   //
+  //   .post(auth, function build(req, res, next) {
+  //     app.carpenter.build({
+  //       data: req.body || {}
+  //     }, function client(error, response) {
+  //       if (error) return next(error);
 
-        response.pipe(res);
-      }).on('error', next);
-    });
+  //       response.pipe(res);
+  //     }).on('error', next);
+  //   });
+
+  app.routes.put('/builds/:pkg/:env?', auth, asynk(async(req, res) => {
+    // Assume no locales are supplied, so just built in en-US, and no artifactory needed
+    // Unpack the base64 string content into a proper directory of code on disk
+    // walk directory to gather files, and publish
+    // create version and package records in system
+    const payload = req.body;
+    const models = app.models;
+    const pkg = models.Package.fromPublish(payload);
+    var version = pkg.version;
+    var name = pkg.name;
+
+    const versionRecord = await manager.getVersionRecord(name, version);
+    if (versionRecord) {
+      // return early by sending error
+      return res.status(403).send({ message: 'version record ddddd already exists for this package' });
+    }
+
+    const rootPath = path.join(tmpdir(), 'pkg');
```

Comment:
this root path specifically could be created on startup in some preboot so we don't have to mkdir it. We might want to use the `mkdirp` package for `mkdir -p` functionality in case it was already created.

Reply:
doesn't mkdirp just call fs.mkdir(path, { recursive: true }) https://www.npmjs.com/package/mkdirp#native-implementation. which is what i am doing in build-manager's writeToPath function with  `await fsPromises.mkdir(unpackPath, { recursive: true })`?  or am I missing something? even if the root path is created on startup, when we unpack each individual file, it is still going to some sub directory of the root path (/unpack/packagename-version/filename), so mkdir still needs to be called?

Reply:
@wjarvis1 ok, I wasn't sure about the behavior, if it doesn't error when it already exists and its just creating the new directories then its probably fine.

Reply:
yeah seems fine -  i tried it locally with an existing directory